File: 015_Binary_Search_Advanced_Problems/Painter_Partition.swift
class Solution {

    func possible(_ arr: [Int], _ day : Int , _ m : Int , _  k : Int) -> Bool {
        var count = 0
        var bouquetCount = 0
        for i in 0..<arr.count {
            if (arr[i] <= day) {
                count += 1
            }
            else {
             bouquetCount += (count / k)
             count = 0
            }
        }
        bouquetCount += (count / k)
        return bouquetCount >= m
    }

    
    func minDays(_ bloomDay: [Int], _ m: Int, _ k: Int) -> Int {
        let value : Int = m * k
        if(value > bloomDay.count) {
            return -1
        }
        let minimum = bloomDay.min() 
        let maximum = bloomDay.max()
        var low = minimum!
        var high = maximum!

        while( low <= high) {
            let mid = (low + high) / 2;
            if (possible(bloomDay,mid,m,k)) {
                high = mid - 1
            }
            else {
                low = mid + 1
            }
        }
        return low

    }
}


let solution = Solution()
// Example input
let bloomDay = [1, 10, 3, 10, 2]
let m = 3
let k = 1

// Call the minDays method
let result = solution.minDays(bloomDay, m, k)

// Print the result
if result == -1 {
    print("It is not possible to make \(m) bouquets with \(k) flowers each.")
} else {
    print("Minimum number of days to make \(m) bouquets with \(k) flowers each: \(result)")
}

File: 015_Binary_Search_Advanced_Problems/Painter_Partition.py
class Solution:
    def possible(self, arr, day, m, k):
        count = 0
        bouquet_count = 0

        for i in range(len(arr)):
            if arr[i] <= day:
                count += 1
            else:
                bouquet_count += count // k
                count = 0

        bouquet_count += count // k
        return bouquet_count >= m

    def min_days(self, bloom_day, m, k):
        value = m * k
        if value > len(bloom_day):
            return -1

        minimum = min(bloom_day)
        maximum = max(bloom_day)
        low = minimum
        high = maximum

        while low <= high:
            mid = (low + high) // 2
            if self.possible(bloom_day, mid, m, k):
                high = mid - 1
            else:
                low = mid + 1

        return low


# Example input
bloom_day = [1, 10, 3, 10, 2]
m = 3
k = 1

# Create an instance of the Solution class
solution = Solution()

# Call the min_days method
result = solution.min_days(bloom_day, m, k)

# Print the result
if result == -1:
    print(f"It is not possible to make {m} bouquets with {k} flowers each.")
else:
    print(f"Minimum number of days to make {m} bouquets with {k} flowers each: {result}")
